File: src-tauri/src/commands/diagnostics.rs
use std::path::PathBuf;
use std::sync::OnceLock;

use sha2::Digest;
use tauri::State;

/// Global session ID set once at app startup, shared across all log calls.
pub static SESSION_ID: OnceLock<String> = OnceLock::new();

pub fn session_id() -> &'static str {
    SESSION_ID.get().map(|s| s.as_str()).unwrap_or("unknown")
}

pub struct DiagnosticsState {
    pub log_path: PathBuf,
}

#[derive(serde::Deserialize)]
pub struct FrontendEvent {
    pub event: String,
    pub payload: serde_json::Value,
}

/// Maps error message strings to safe, non-private error kind labels.
pub fn classify_error(msg: &str) -> &'static str {
    if msg.contains("not found") || msg.contains("Repository not found") {
        "not_found"
    } else if msg.contains("permission") || msg.contains("Permission") {
        "permission_denied"
    } else if msg.contains("No repository") {
        "no_repo_open"
    } else if msg.contains("No active session") {
        "no_session"
    } else if msg.contains("Bare repository") {
        "bare_repo"
    } else if msg.contains("No commits selected") {
        "no_commits_selected"
    } else {
        "unknown"
    }
}

/// Hashes an input string to an 8-character lowercase hex string (first 4 bytes of SHA-256).
/// Used to correlate events from the same repo without logging the path.
pub fn hash_string(input: &str) -> String {
    let mut hasher = sha2::Sha256::new();
    hasher.update(input.as_bytes());
    let result = hasher.finalize();
    hex::encode(&result[..4])
}

/// Returns the set of field names that are safe to log for a given event type.
/// Any field not in this list is silently dropped by `sanitize`.
fn allowed_fields(event: &str) -> &'static [&'static str] {
    match event {
        "repo_open_attempt" => &[],
        "repo_open_success" => &["hashed_repo_id", "commit_count", "duration_ms"],
        "repo_open_failure" => &["error_kind"],
        "commits_loaded" => &["count", "duration_ms"],
        "diff_loaded" => &["selected_commit_count", "file_count", "total_changed_lines", "duration_ms"],
        "file_selected" => &["count"],
        "session_created" => &[],
        "session_loaded" => &[],
        "session_ended" => &["comment_count", "edit_count", "duration_ms"],
        "session_exported" => &[],
        "session_abandoned" => &["comment_count", "edit_count"],
        "comment_added" => &["severity", "comment_type"],
        "comment_resolved" => &[],
        "comment_deleted" => &[],
        "edit_applied" => &[],
        "ipc_error" => &["command_name", "error_kind"],
        "js_error" => &["error_class", "component_stack_depth"],
        "app_context" => &["screen_width", "screen_height", "locale"],
        "updater_check" => &["available", "current_version", "latest_version"],
        "updater_error" => &["error_kind"],
        _ => &[],
    }
}

/// Returns true if a string value is safe to log (no path separators, not too long).
fn is_safe_string(s: &str) -> bool {
    if s.len() > 64 {
        return false;
    }
    // Heuristic: strings with path separators are likely file paths
    if s.contains('/') || s.contains('\\') {
        return false;
    }
    true
}

/// Sanitizes a frontend event payload, enforcing the field allowlist and
/// redacting any string values that look like file paths.
pub fn sanitize(event: &str, payload: serde_json::Value) -> serde_json::Value {
    let allowed = allowed_fields(event);
    let obj = match payload {
        serde_json::Value::Object(map) => map,
        _ => return serde_json::Value::Object(serde_json::Map::new()),
    };

    let mut result = serde_json::Map::new();
    let mut redacted = false;

    for (key, value) in obj {
        if !allowed.contains(&key.as_str()) {
            continue;
        }
        match &value {
            serde_json::Value::String(s) => {
                if is_safe_string(s) {
                    result.insert(key, value);
                } else {
                    result.insert(key, serde_json::Value::String("[REDACTED]".to_string()));
                    redacted = true;
                }
            }
            serde_json::Value::Number(_) | serde_json::Value::Bool(_) | serde_json::Value::Null => {
                result.insert(key, value);
            }
            // Drop arrays and nested objects — not expected in any event payload
            _ => {}
        }
    }

    if redacted {
        tracing::warn!(
            event = "sanitizer_redacted",
            original_event = %event,
            session_id = %session_id(),
        );
    }

    serde_json::Value::Object(result)
}

// ── IPC Commands ─────────────────────────────────────────────────────

/// Accepts a structured event from the frontend, sanitizes it, and writes it
/// to the log file via the tracing infrastructure.
#[tauri::command]
pub fn log_event(
    event: FrontendEvent,
    _state: State<'_, DiagnosticsState>,
) -> Result<(), String> {
    let clean_payload = sanitize(&event.event, event.payload);
    tracing::info!(
        event = %event.event,
        payload = %clean_payload,
        session_id = %session_id(),
        source = "frontend",
    );
    Ok(())
}

/// Returns the path to the log directory so the frontend can open it for the user.
#[tauri::command]
pub fn get_log_path(state: State<'_, DiagnosticsState>) -> Result<String, String> {
    Ok(state.log_path.to_string_lossy().to_string())
}

/// Hashes an input string to an 8-char hex ID. Used by the frontend to hash
/// repo paths before logging them, ensuring no raw paths appear in logs.
#[tauri::command]
pub fn hash_string_cmd(input: String) -> Result<String, String> {
    Ok(hash_string(&input))
}

File: src/diagnostics.ts
import { invoke } from "@tauri-apps/api/core";

// ── Payload type ────────────────────────────────────────────────────
// Intentionally narrow: only primitive values allowed.
// This prevents accidentally passing objects that might contain private data.
type SafeValue = string | number | boolean | null;
type DiagnosticPayload = Record<string, SafeValue>;

// ── Internal helpers ────────────────────────────────────────────────

function logEvent(event: string, payload: DiagnosticPayload = {}): void {
	// Fire-and-forget: logging must never throw or block the caller.
	invoke("log_event", { event: { event, payload } }).catch(() => {});
}

/** Returns a function that returns elapsed milliseconds since it was called. */
export function startTimer(): () => number {
	const t0 = performance.now();
	return () => Math.round(performance.now() - t0);
}

/** Cache of repo path → 8-char hex hash, populated lazily. */
const hashCache = new Map<string, string>();

/**
 * Hashes a string (e.g. a repo path) to an 8-char hex ID using the Rust backend.
 * The result is cached so subsequent calls for the same input are free.
 * NEVER log raw paths — always pass through this function first.
 */
export async function hashPath(path: string): Promise<string> {
	const cached = hashCache.get(path);
	if (cached) return cached;
	try {
		const hashed = await invoke<string>("hash_string_cmd", { input: path });
		hashCache.set(path, hashed);
		return hashed;
	} catch {
		return "unknown";
	}
}

/**
 * Maps an unknown error to a safe, non-private label for logging.
 * NEVER log e.message or e.toString() directly — those strings often contain file paths.
 */
export function classifyFrontendError(e: unknown): string {
	if (e instanceof Error) {
		const msg = e.message.toLowerCase();
		if (msg.includes("not found")) return "not_found";
		if (msg.includes("permission")) return "permission_denied";
		if (msg.includes("no repository")) return "no_repo_open";
		if (msg.includes("no active session")) return "no_session";
		if (msg.includes("bare repository")) return "bare_repo";
		if (msg.includes("no commits")) return "no_commits_selected";
	}
	return "unknown";
}

// ── Public API ──────────────────────────────────────────────────────
// Each method is a typed wrapper that enforces the safe-fields contract.
// Callers cannot accidentally pass private data — the parameters are typed
// to only accept values that are safe to log.

export const diag = {
	appContext(screenWidth: number, screenHeight: number, locale: string): void {
		logEvent("app_context", { screen_width: screenWidth, screen_height: screenHeight, locale });
	},

	repoOpenAttempt(): void {
		logEvent("repo_open_attempt");
	},

	repoOpenSuccess(hashedRepoId: string, commitCount: number, durationMs: number): void {
		logEvent("repo_open_success", { hashed_repo_id: hashedRepoId, commit_count: commitCount, duration_ms: durationMs });
	},

	repoOpenFailure(errorKind: string): void {
		logEvent("repo_open_failure", { error_kind: errorKind });
	},

	commitsLoaded(count: number, durationMs: number): void {
		logEvent("commits_loaded", { count, duration_ms: durationMs });
	},

	diffLoaded(commitCount: number, fileCount: number, totalLines: number, durationMs: number): void {
		logEvent("diff_loaded", {
			selected_commit_count: commitCount,
			file_count: fileCount,
			total_changed_lines: totalLines,
			duration_ms: durationMs,
		});
	},

	fileSelected(count: number): void {
		logEvent("file_selected", { count });
	},

	sessionCreated(): void {
		logEvent("session_created");
	},

	sessionLoaded(): void {
		logEvent("session_loaded");
	},

	sessionEnded(commentCount: number, editCount: number, durationMs: number): void {
		logEvent("session_ended", { comment_count: commentCount, edit_count: editCount, duration_ms: durationMs });
	},

	sessionAbandoned(commentCount: number, editCount: number): void {
		logEvent("session_abandoned", { comment_count: commentCount, edit_count: editCount });
	},

	sessionExported(): void {
		logEvent("session_exported");
	},

	commentAdded(severity: string, commentType: string): void {
		logEvent("comment_added", { severity, comment_type: commentType });
	},

	commentResolved(): void {
		logEvent("comment_resolved");
	},

	commentDeleted(): void {
		logEvent("comment_deleted");
	},

	editApplied(success: boolean): void {
		logEvent("edit_applied", { success });
	},

	ipcError(commandName: string, errorKind: string): void {
		logEvent("ipc_error", { command_name: commandName, error_kind: errorKind });
	},

	jsError(errorClass: string, componentStackDepth: number): void {
		logEvent("js_error", { error_class: errorClass, component_stack_depth: componentStackDepth });
	},

	updaterCheck(available: boolean, currentVersion: string | null, latestVersion: string | null): void {
		logEvent("updater_check", {
			available,
			current_version: currentVersion,
			latest_version: latestVersion,
		});
	},

	updaterError(errorKind: string): void {
		logEvent("updater_error", { error_kind: errorKind });
	},

	bugReportSubmitted(success: boolean): void {
		logEvent("bug_report_submitted", { success });
	},
};

/** Maps updater plugin errors to safe, non-private kind labels. */
export function classifyUpdaterError(e: unknown): string {
	if (e instanceof Error || typeof e === "string") {
		const msg = (e instanceof Error ? e.message : e).toLowerCase();
		if (msg.includes("signature")) return "signature_failed";
		if (msg.includes("no available") || msg.includes("target")) return "no_platform";
		if (msg.includes("parse") || msg.includes("json")) return "parse_error";
		if (msg.includes("network") || msg.includes("fetch") || msg.includes("request") || msg.includes("connect") || msg.includes("dns")) return "network";
	}
	return "unknown";
}
